File: PookieBear-iOS/PookieBear/Models/Activity.swift
import Foundation
import CoreLocation

struct Activity: Codable, Identifiable {
    let id: String
    let dogId: String
    let activityType: ActivityType
    let startTime: Date
    let endTime: Date?
    let durationSeconds: Int?
    let distanceMeters: Double?
    let averagePace: Double?
    let routePolyline: String?
    let routeCoordinates: [Coordinate]?
    let caloriesBurned: Int?
    let steps: Int?
    let temperatureFahrenheit: Double?
    let weatherConditions: [String: String]?
    let notes: String?
    let photos: [String]?
    let intensity: ActivityIntensity?
    let location: String?
    let createdAt: Date

    enum ActivityType: String, Codable {
        case walk, playtime, training, swimming, fetch, park_visit, daycare, hiking

        var displayName: String {
            switch self {
            case .walk: return "Walk"
            case .playtime: return "Playtime"
            case .training: return "Training"
            case .swimming: return "Swimming"
            case .fetch: return "Fetch"
            case .park_visit: return "Park Visit"
            case .daycare: return "Daycare"
            case .hiking: return "Hiking"
            }
        }

        var icon: String {
            switch self {
            case .walk: return "figure.walk"
            case .playtime: return "tennis.racket"
            case .training: return "brain.head.profile"
            case .swimming: return "figure.pool.swim"
            case .fetch: return "tennisball"
            case .park_visit: return "tree"
            case .daycare: return "building.2"
            case .hiking: return "mountain.2"
            }
        }
    }

    enum ActivityIntensity: String, Codable {
        case low, moderate, high
    }

    struct Coordinate: Codable {
        let latitude: Double
        let longitude: Double

        var clLocation: CLLocationCoordinate2D {
            CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
        }
    }

    var distanceMiles: Double? {
        guard let meters = distanceMeters else { return nil }
        return meters / 1609.34
    }

    var durationMinutes: Int? {
        guard let seconds = durationSeconds else { return nil }
        return seconds / 60
    }

    var durationFormatted: String {
        guard let seconds = durationSeconds else { return "N/A" }
        let hours = seconds / 3600
        let minutes = (seconds % 3600) / 60
        if hours > 0 {
            return "\(hours)h \(minutes)m"
        }
        return "\(minutes) min"
    }
}

struct ActivityListResponse: Codable {
    let activities: [Activity]
}

struct ActivityStats: Codable {
    let period: String
    let stats: Stats

    struct Stats: Codable {
        let totalActivities: Int
        let totalDistanceMeters: Double
        let totalDistanceMiles: Double
        let totalDurationMinutes: Int
        let totalCalories: Int
        let walkCount: Int
        let averageWalkDistanceMeters: Double
        let averageWalkDistanceMiles: Double
    }
}

File: pookiebear-backend/src/entities/Activity.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { Dog } from './Dog';
import { User } from './User';

export enum ActivityType {
  WALK = 'walk',
  PLAYTIME = 'playtime',
  TRAINING = 'training',
  SWIMMING = 'swimming',
  FETCH = 'fetch',
  PARK_VISIT = 'park_visit',
  DAYCARE = 'daycare',
  HIKING = 'hiking',
}

export enum ActivityIntensity {
  LOW = 'low',
  MODERATE = 'moderate',
  HIGH = 'high',
}

@Entity('activities')
export class Activity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ type: 'uuid' })
  dogId: string;

  @ManyToOne(() => Dog, (dog) => dog.activities, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'dogId' })
  dog: Dog;

  @Column({ type: 'uuid' })
  userId: string;

  @ManyToOne(() => User)
  @JoinColumn({ name: 'userId' })
  user: User;

  @Column({ type: 'enum', enum: ActivityType })
  activityType: ActivityType;

  @Column({ type: 'timestamp' })
  startTime: Date;

  @Column({ type: 'timestamp', nullable: true })
  endTime: Date;

  @Column({ type: 'int', nullable: true })
  durationSeconds: number;

  @Column({ type: 'float', nullable: true })
  distanceMeters: number;

  @Column({ type: 'float', nullable: true })
  averagePace: number;

  @Column({ type: 'text', nullable: true })
  routePolyline: string;

  @Column({ type: 'jsonb', nullable: true })
  routeCoordinates: object[];

  @Column({ type: 'int', nullable: true })
  caloriesBurned: number;

  @Column({ type: 'int', nullable: true })
  steps: number;

  @Column({ type: 'jsonb', nullable: true })
  pauses: object[];

  @Column({ type: 'jsonb', nullable: true })
  weatherConditions: object;

  @Column({ type: 'float', nullable: true })
  temperatureFahrenheit: number;

  @Column({ type: 'text', nullable: true })
  notes: string;

  @Column({ type: 'simple-array', nullable: true })
  photos: string[];

  @Column({ type: 'enum', enum: ActivityIntensity, nullable: true })
  intensity: ActivityIntensity;

  @Column({ type: 'varchar', nullable: true })
  location: string;

  @Column({ type: 'boolean', default: true })
  synced: boolean;

  @CreateDateColumn()
  createdAt: Date;
}